### DISCUSSION TITLE
Replace file menu item

### DISCUSSION THREAD

**Discussion by 6elkir** [rejetto/hfs#303] [Q&A] [answered]
Hello, I know how to add a new item to the file menu. But now there was a request to REPLACE the menu item with your own. In particular, as many people already know from the questions, I have improved the writing option for viewing files and I need to replace the basic "Show" with the opening of the standard viewer with my own. Please tell me how to do it?

Adding item to file menu for example
`HFS . onEvent
    (
        'fileMenu',
        function ( { entry } )
        {
            if ( typeof mmp_cfg . stotage . links [entry . uri] !== 'number' )
            {
                return;
            }
            return {
                label   : 'View file',
                icon    : 'doc' ,
                onClick : () => dnr_view_entry ( entry )
            };
        }			
    );`

**Comment by rejetto** [comment]
I will try to make the documentation clearer about it.
You not only get "entry", you also get "menu".
So, if "index" is the position of the item you want to replace, you can just do

menu[index] = { label: "your feature".... }

and return nothing (otherwise you would be replacing AND adding, you'll see it twice).

Anyway, it's a problem to find WHAT item is the show, because the text could be translated.
I can solve this problem in next release.

**Comment by rejetto** [comment]
in the meantime, you can use this non-perfect method to find the index of the "show"
`const index = menu.findIndex(x => x.icon === 'image')`

so, the full text  would be like

**Comment by 6elkir** [comment]
> I will try to make the documentation clearer about it. You not only get "entry", you also get "menu". So, if "index" is the position of the item you want to replace, you can just do
> 
> menu[index] = { label: "your feature".... }
> 
> and return nothing (otherwise you would be replacing AND adding, you'll see it twice).
> 
> Anyway, it's a problem to find WHAT item is the show, because the text could be translated. I can solve this problem in next release.

Yes it would be cool! How the event handler passes the 2nd parameter, the maximum allowed, object / array element, so that the menu can be iterated and found the same one)

**Comment by rejetto** [comment, marked as answer]
the "menu" is already passed in current version. This is a more complete example:
```js
HFS.onEvent('fileMenu', ({ entry, menu }) => {
  const index = menu.findIndex(x => x.icon === 'image')
  menu[index] = { label: "your feature" }
})
```

i didn't test it

**Reply by 6elkir** [reply]
Great! Thanks for the help! I think the documentation needs to be brought to mind and then the community will finish the feature right away)))
Here's what happened now:
https://share.dvor-net.ptz.ru/t2fHlG9Yft.mp4

Code:
```js
HFS . onEvent
    (
        'fileMenu',
        function ( { entry, menu } )
        {
            if ( typeof mmp_cfg . stotage . links [entry . uri] !== 'number' )
            {
                return;
            }
            
            var callback = () => dnr_view_entry ( entry );
            
            var item = HFS . _ . find
            (
                menu,
                {
                    'icon'  : 'image',
                    'label' : 'Show',
                }
            );
            
            if ( item !== undefined )
            {
                item . onClick = callback;
                
                return;
            }
            
            
            return {
                label   : 'Show',
                icon    : 'image' ,
                onClick : callback
            };
        }			
    );
```

**Comment by rejetto** [comment]
cool, but don't include "label: show" when you *find*, otherwise it won't work when it is translated to another language.
in next version I will add an "id" field, so you can search with it, much safer.

**Reply by 6elkir** [reply]
Okay)
